### PR TITLE
Dynamic image resampling and plotting 2D coordinates

### PR DESCRIPTION
**Pull request by nvaytet**
We make heavy use of `rebin` in the 2D image plots to:
- resample very large images
- plot data with 2D coordinates.

Details of the implementation are a little difficult to follow. I recommend a walkthrough for the reviewer.
Note that the data is now **always** rebinned to display the image, no matter what the resolution is.
Checking on resolution was creating too many checks and if statements.

**Note** that when using 2D coords, data outside of the range for each row is not masked, it is simply zero (output from rebin). This is different from Mantid.

**A few examples to try out:**

1. A large image: should by dynamically resampling with zoom. Transpose X/Y buttons should also work.
```Py
N = 6000
M = 3000
xx = np.arange(N, dtype=np.float64)
yy = np.arange(M, dtype=np.float64)
x, y = np.meshgrid(xx, yy)
b = N/500.0
c = M/100.0
r = np.sqrt(((x-c)/b)**2 + (y/b)**2)
a = np.sin(r) + np.random.random([M, N])
d = sc.Dataset()
d.coords['x'] = sc.Variable(['x'], values=xx, unit=sc.units.m)
d.coords['y'] = sc.Variable(['y'], values=yy, unit=sc.units.m)
d['a'] = sc.Variable(['y', 'x'], values=a, unit=sc.units.counts)
plot(d)
```
Screenshot at 2020-07-30 22-00-28

2. A 2D dataset with a 2D coordinate and masks:
```Py
N = 10
M = 5
x = np.arange(N+1).astype(np.float64)
y = np.arange(M).astype(np.float64)
xx, yy = np.meshgrid(x, y)
z = np.random.random([M, N])
for i in range(M):
    xx[i] *= (i + 1.0)
d = sc.Dataset()
d.coords['x'] = sc.Variable(['y', 'x'], values=xx, unit=sc.units.m)
d.coords['y'] = sc.Variable(['y'], values=y, unit=sc.units.m)
d['a'] = sc.Variable(['y', 'x'], values=z, unit=sc.units.counts)
d.masks['b'] = sc.Variable(['y', 'x'], values=np.where(z<0.5, True, False), dtype=bool)
plot(d)
```
Screenshot at 2020-07-30 22-06-19

3. A large image with a 2D coordinate. The use of rebin means that we can zoom at 'infinite' resolution on the very small bins in the lower left corner.
```Py
N = 600
M = 300
x = np.arange(N+1).astype(np.float64)
y = np.arange(M).astype(np.float64)
xx, yy = np.meshgrid(x, y)
z = np.random.random([M, N])
for i in range(M):
    xx[i] *= (i + 1.0)**2
d = sc.Dataset()
d.coords['x'] = sc.Variable(['y', 'x'], values=xx, unit=sc.units.m)
d.coords['y'] = sc.Variable(['y'], values=y, unit=sc.units.m)
d['a'] = sc.Variable(['y', 'x'], values=z, unit=sc.units.counts)
plot(d)
```
Screenshot at 2020-07-30 22-08-24

4. Compare figure loading times between Mantid and this implementation in the neutron notebook:
(also not sure what is going on with Mantid's colorbar range)
Screenshot at 2020-07-30 22-53-28

**Various other things to check:**

- Use labels/non-dimension coord instead of dimension coord (using `axes=[...]`)
- Have a look at the panning behaviour: it waits for a large enough change to the axis limits before rebinning, to avoid too many updates while panning. This yields white borders on the side of the image while panning. Is this acceptable? I don't know how much panning is used to inspect data (not so much in my experience). On that note, the Mantid display for large images does not have the white borders, but has some flickering in the image (from different rows being selected by the strides) which we do not have. I don't know which is better,
- Check behaviour with realigned data? (I did a quick test but did not play with it extensively)

### CODE REVIEW

File: python/src/scipp/plot/plot_2d.py [hunk 1]
```diff
@@ -182,72 +192,83 @@ def update_buttons(self, owner, event, dummy):
...
+                    xory = dims[iother]
+                    for i in range(shp[iother]):
+                        edges[xory, i] = centers_to_edges(vcoord[xory,
+                                                                 i].values)
+                else:
+                    edges = sc.Variable(
+                        dims,
+                        values=centers_to_edges(
+                            self.slider_coord[self.name][param["dim"]].values),
+                        dtype=sc.dtype.float64)
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:266]: Why is special handling for 2D coords required here? What is different?

File: python/src/scipp/plot/plot_2d.py [hunk 2]
```diff
@@ -256,111 +277,222 @@ def update_axes(self):
...
+            self.xyrebin["y"].dims[0]: self.xyrebin["y"],
+        },
+                           data=sc.Variable(dims=self.button_dims,
+                                            values=np.ones([
+                                                self.xyrebin["y"].shape[0] - 1,
+                                                self.xyrebin["x"].shape[0] - 1
+                                            ]),
+                                            dtype=dslice.dtype,
+                                            unit=sc.units.one))
+        arr *= dslice
```
Comment (SimonHeybrock) [inline python/src/scipp/plot/plot_2d.py:482]: Easier to use https://scipp.github.io/generated/scipp.transpose.html#scipp.transpose by hand?

Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:482]: yes! much easier.
This is a recent addition, right?

Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:482]: Has existed for a long time, but Python bindings where missing, Owen added them earlier this week.

File: python/src/scipp/plot/plot_2d.py (hunk 1 above)
Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:266]: So here I convert each row from center to edges.
The function `centers_to_edges` assumes it's getting a 1D array, so i can't send it the `.values` directly.

Comment (SimonHeybrock) [reply python/src/scipp/plot/plot_2d.py:266]: Wouldn't it make more sense to fix `centers_to_edges`? Shouldn't be hard to write it for N-D data?

Comment (nvaytet) [reply python/src/scipp/plot/plot_2d.py:266]: yes, would probably make more sense to make `centers_to_edges` handle more dimensions.
Not sure it's super easy to write though...

**Comment by nvaytet** [comment]
> 1. Swapping axes resets the zoom to home. Is this intentional?
Yes, and this has always been the case. It's the simplest way to do things and I think it's fine for now, until someone requests otherwise. It's also not super obvious what to do otherwise. If you had zoomed in to the upper left corner, do you then move to the lower right corner, or do you stay in the upper left corner?

**Comment by SimonHeybrock** [comment]
> > 1. Swapping axes resets the zoom to home. Is this intentional?
> 
> Yes, and this has always been the case. It's the simplest way to do things and I think it's fine for now, until someone requests otherwise. It's also not super obvious what to do otherwise. If you had zoomed in to the upper left corner, do you then move to the lower right corner, or do you stay in the upper left corner?

I would keep exactly the same viewport, i.e., essentially just transpose the image that is displayed.

**Comment by nvaytet** [comment]
> 2. There is a small shift(?) between the mask and data. See thin white line around mask:

i think this may be an effect of the `rebin` at the bin edges, which blurs the boundaries of pixels (in this case it's averaging with a nan which makes it white?). Sometimes I see it and sometimes I don't.
Is it less visible if you increase the resolution? e.g. `plot(d, resolution=1024)`?
If so, then I think that might explain it.

**Comment by nvaytet** [comment]
> 3. Plotting 3d data does not work (see exception). I tried swapping axes so the ragged dimension is not the inner dim but the out (slider) dim. The slider does not show. My data has a 3d ragged coord:

I don't really handle ND coords on sliders at the moment....
and I haven't even thought about ND coords for the 3d plotting...
Maybe it's ok to defer to another PR? This was mainly for 2d plots.

**Comment by nvaytet** [comment]
Code example for 3D corodinate:
```Py
N = 10
M = 8
L = 5
x = np.arange(N+1).astype(np.float64)
y = np.arange(M).astype(np.float64)
z = np.arange(L).astype(np.float64)
zz, yy, xx = np.meshgrid(z, y, x, indexing='ij')
a = np.random.random([L, M, N])
for i in range(M):
    for j in range(L):
        xx[j, i, :] *= (i + j + 1.0)
d = sc.Dataset()
d.coords['x'] = sc.Variable(['z', 'y', 'x'], values=xx, unit=sc.units.m)
d.coords['y'] = sc.Variable(['y'], values=y, unit=sc.units.m)
d.coords['z'] = sc.Variable(['z'], values=z, unit=sc.units.m)
d['a'] = sc.Variable(['z', 'y', 'x'], values=a, unit=sc.units.counts)
plot(d)
```

**Comment by nvaytet** [comment]
When a ND coordinate is on a slider, the label becomes `slice-17`, wher `17` is the index in the coordinate array, as there is no longer a single coordinate value for the current slider position/index.

**Pull request merged**